Have es2015 rest transform safely use `arguments`

vhf opened the pull request:
Hi, I audited some code generated by Babel and noticed that in some cases some improvements could be made to ensure that V8 optimiser (Crankshaft) will not bail out a little too soon. I focused on the use of `arguments`: using `arguments` in an unsafe way prevents V8 (and therefore Node.js) to optimise a function. Here is a first patch in the direction of using `arguments` safely in Babel es2015-transformed code. Although far from perfect in terms of implementation (first time contributor here, still learning Babel internals, happy to receive your comments and/or to chat on slack or by email) this patch will probably give an idea of what could be done, and could be followed by other patches aiming towards the goal of avoiding generation of V8-unoptimisable code.

With this PR I'm only focusing on the very first and easiest part of my findings. In the future, I'll try to fix some other unsafe uses of `arguments` in every transformation and then try to focus on other possible V8 optimisations.

From what I understood, performances are more important to Babel than prettiness. This patch somewhat impacts readability of es2015-presets generated code. But it also has impact on performance.

Although tests are green and I did my best to adhere to what I think are Babel's codebase best practices, please review very carefully. For example, I was not sure whether to include `optimiseLoadStatement` definition in `visitor` (as seen in `defaults.js`) or to define them outside of `visitor`'s scope.

sebmck commented on packages/babel-plugin-transform-es2015-parameters/src/rest.js:
```diff
@@ -154,7 +166,13 @@ export let visitor = {
-            optimiseMemberExpression(candidate.parent, state.offset);
+            if (t.isReturnStatement(candidate.parentPath.parent)
+              || t.isIdentifier(candidate.parentPath.parent.id)) {
+              let optimized = optimiseLoadStatement(candidate.parent, state.offset, argsId);
```
`optimized` -> `optimised`.

sebmck commented on packages/babel-plugin-transform-es2015-parameters/src/rest.js:
```diff
@@ -154,7 +166,13 @@ export let visitor = {
           candidate.replaceWith(argsId);
           if (candidate.parentPath.isMemberExpression()) {
-            optimiseMemberExpression(candidate.parent, state.offset);
+            if (t.isReturnStatement(candidate.parentPath.parent)
```
There's way too much nesting here, can you abstract it out into a function or something?

sebmck commented on packages/babel-plugin-transform-es2015-parameters/src/rest.js:
```diff
@@ -116,7 +120,15 @@ export let visitor = {
 
-    // support patterns
+    function optimiseLoadStatement(parent, offset) {
+      let newExpr = loadRest({
```
no need to assign it to a variable to only return it in the next statement, merge this into a single statement pelase

vhf commented:
Thanks for your comments. In this first commit I fixed the `optimized` typo, removed the useless `newExpr` temp variable and reduced nesting by factoring out a function. This is almost guaranteed to work.

The second commit however is more problematic: while trying to understand how this transformation works, I noticed that removing parts of it did not change anything to the tests status. I tried to think of cases where this transformation would not behave as expected but did not come up with anything not already covered in the test suite, so I included this commit. I'm very genuinely interested by the history of these parts of code, although I did not dig past "v6.0.0 revolution" in this file's history.

thejameskyle commented on packages/babel-core/test/fixtures/transformation/es6.parameters/rest-member-expression-optimisation/expected.js:
```diff
@@ -1,9 +1,9 @@
 var t = function () {
-    var x = arguments[0];
-    var y = arguments[1];
+    var x = arguments.length <= 0 || arguments[0] === undefined ? undefined : arguments[0];
```
Is the `|| arguments[0] === undefined` necessary for the v8 optimisation? It seems odd that it would

vhf replied:
You're right, it's not. In most cases `arguments.length <= 0 || (undefined || arguments[0])` (I suppose that's what you suggest?) behaves the same way, but they are not really equivalent. Consider the following:

```javascript
var _arguments = {length: 0};
console.log(_arguments.length <= 0 || _arguments[0] === undefined ? undefined : _arguments[0]);
console.log(_arguments.length <= 0 || (undefined || _arguments[0]));
```
First case will return `undefined`, second case will return `true`.

vhf commented:
I finally fixed my PR. Tests are green, coverage remains unchanged.

thejameskyle commented:
Looks good to me, @sebmck ?

sebmck commented:
Super disappointing that VMs require this :( I'll hopefully get around to reviewing this early next week. Please feel free to ping this issue if I forget. Really sorry for the delay!

eventualbuddha commented on packages/babel-plugin-transform-es2015-parameters/src/rest.js:
```diff
@@ -11,6 +11,10 @@ let buildRest = template(`
 `);
 
+let loadRest = template(`
+  ARGUMENTS.length <= KEY || ARGUMENTS[KEY] === undefined ? undefined : ARGUMENTS[KEY]
```
Minor style issue: would `INDEX` be more appropriate than `KEY`?

vhf replied:
I don't think so. I chose `KEY` because it was consistent with the existing codebase, e.g. [0](https://github.com/babel/babel/blob/master/packages/babel-plugin-transform-es2015-parameters/src/default.js), [1](https://github.com/babel/babel/blob/master/packages/babel-plugin-transform-es2015-for-of/src/index.js), [2](https://github.com/babel/babel/blob/master/packages/babel-plugin-transform-es2015-parameters/src/rest.js)

thejameskyle replied:
I would've picked INDEX, but it doesn't really matter either way.

vhf commented:
Ping @sebmck :)

sebmck commented:
Thanks for doing this @vhf! Extremely sorry for the delay in merging this.

kuraga commented on packages/babel-plugin-transform-es2015-parameters/test/fixtures/parameters/rest-member-expression-optimisation/expected.js:
```diff
@@ -1,9 +1,9 @@
 var t = function () {
-    var x = arguments[0];
-    var y = arguments[1];
+    var x = arguments.length <= 0 || arguments[0] === undefined ? undefined : arguments[0];
```
@vhf Why not `arguments.length <= 0 ? undefined : arguments[0];`?

vhf replied:
Good catch! The reason behind this code is [this code](https://github.com/babel/babel/blob/master/packages/babel-plugin-transform-es2015-parameters/src/default.js#L8-L9). But you're absolutely right, it's uselessly complicated here.

chpio commented:
my *KEY* gets transformed to *NaN*:

```javascript
	for (let i = 0; i < sources.length; i += 1) {
		const source = sources[i];
		// ...
	}
```

```javascript
	for (var i = 0; i < arguments.length; i += 1) {
		var source = arguments.length <= NaN || arguments[NaN] === undefined ? undefined : arguments[NaN];
		// ...
	}
```

vhf commented:
Should be 

```js
function func() {
  for (var i = 0; i < arguments.length; i += 1) {
    var source = arguments[i];
    // ...
  }
}
```

instead.

Fixed by #3165
